fish-rust/src/output.rs

```diff
@@ -8,12 +24,706 @@
     }
 }
 
-pub fn output_set_color_support(value: ColorSupport) {
-    extern "C" {
-        pub fn output_set_color_support(value: libc::c_int);
+/// Whether term256 and term24bit are supported.
+static COLOR_SUPPORT: AtomicU8 = AtomicU8::new(0);
+
+/// FFI bits.
+#[no_mangle]
+extern "C" fn output_get_color_support() -> u8 {
+    COLOR_SUPPORT.load(Ordering::Relaxed)
+}
+
+#[no_mangle]
+extern "C" fn output_set_color_support(val: u8) {
+    COLOR_SUPPORT.store(val, Ordering::Relaxed);
+}
+
+/// Returns true if we think tparm can handle outputting a color index.
+fn term_supports_color_natively(c: i32, term: &Term) -> bool {
+    #[allow(clippy::int_plus_one)]
+    if let Some(max_colors) = term.max_colors {
+        max_colors >= c + 1
+    } else {
+        false
+    }
+}
+
+pub fn get_color_support() -> ColorSupport {
+    let val = COLOR_SUPPORT.load(Ordering::Relaxed);
+    ColorSupport::from_bits_truncate(val)
+}
+
+pub fn set_color_support(val: ColorSupport) {
+    COLOR_SUPPORT.store(val.bits(), Ordering::Relaxed);
+}
+
+// These are historic. writembs() attempts to write a multibyte string of type &Option<CString> to the given outputter.
+// writembs() will noisily fail, writembs_nofail will not.
+macro_rules! writembs(
+    ($outp: expr, $mbs: expr) => {
+        crate::output::writembs_check($outp, $mbs, std::stringify!($mbs), true, std::file!(), std::line!())
+    }
+);
+
+macro_rules! writembs_nofail(
+    ($outp: expr, $mbs: expr) => {
+        crate::output::writembs_check($outp, $mbs, std::stringify!($mbs), false, std::file!(), std::line!())
+    }
+);
+pub(crate) use writembs_nofail;
+
+fn index_for_color(c: RgbColor) -> u8 {
+    if c.is_named() || !(get_color_support().contains(ColorSupport::TERM_256COLOR)) {
+        return c.to_name_index();
+    }
+    c.to_term256_index()
+}
+
+fn write_color_escape(outp: &mut Outputter, term: &Term, todo: &CStr, mut idx: u8, is_fg: bool) {
+    if term_supports_color_natively(idx.into(), term) {
+        // Use tparm to emit color escape.
+        writembs!(outp, tparm1(todo, idx.into()));
+    } else {
+        // We are attempting to bypass the term here. Generate the ANSI escape sequence ourself.
+        if idx < 16 {
+            // this allows the non-bright color to happen instead of no color working at all when a
+            // bright is attempted when only colors 0-7 are supported.
+            //
+            // TODO: enter bold mode in builtin_set_color in the same circumstance- doing that combined
+            // with what we do here, will make the brights actually work for virtual consoles/ancient
+            // emulators.
+            if term.max_colors == Some(8) && idx > 8 {
+                idx -= 8;
+            }
+            write!(
+                outp,
+                "\x1B[{}m",
+                (if idx > 7 { 82 } else { 30 }) + idx as i32 + ((!is_fg as i32) * 10)
```

Comment:
I consider every use of `as` for arithmetic conversion suspicious, I prefer `i32::from`

Reply:
Good tip, fixed